arbor/fvm_layout.cpp

```diff
@@ -124,43 +126,52 @@ pw_elements<U> pw_over_cable(const mcable_map<T>& mm, mcable cable, U dflt_value
 
 // Construct cv_geometry for cell from locset describing CV boundary points.
 cv_geometry cv_geometry_from_ends(const cable_cell& cell, const locset& lset) {
-    struct mloc_hash {
-        std::size_t operator()(const mlocation& loc) const { return loc.branch ^ std::hash<double>()(loc.pos); }
-    };
-    using mlocation_map = std::unordered_map<mlocation, fvm_size_type, mloc_hash>;
     auto pop = [](auto& vec) { auto h = vec.back(); return vec.pop_back(), h; };
 
     cv_geometry geom;
     const auto& mp = cell.provider();
     const auto& m = mp.morphology();
 
-    if (mp.morphology().empty()) {
+    if (m.empty()) {
         geom.cell_cv_divs = {0, 0};
         return geom;
     }
 
-    auto canon =    [&m](mlocation loc)  { return canonical(m, loc); };
-    auto origin =   [&canon](mcable cab) { return canon(mlocation{cab.branch, cab.prox_pos}); };
-    auto terminus = [&canon](mcable cab) { return canon(mlocation{cab.branch, cab.dist_pos}); };
-
     mlocation_list locs = thingify(lset, mp);
 
-    std::vector<msize_t> n_cv_cables;
-    std::vector<mlocation> next_cv_head;
-    next_cv_head.push_back({mnpos, 0});
+    // Filter out root, terminal locations and repeated locations so as to
+    // avoid trivial CVs outside of fork points. (This is not necessary for
+    // correctness, but is for the convenience of specification by lset.)
 
-    mcable_list cables, all_cables;
-    std::vector<msize_t> branches;
+    auto neither_root_nor_terminal = [&m](mlocation x) {
+        return !(x.pos==0 && x.branch==(m.branch_children(mnpos).size()>1u? mnpos: 0)) // root?
+            && !(x.pos==1 && m.branch_children(x.branch).empty()); // terminal?
+    };
+    locs.erase(std::partition(locs.begin(), locs.end(), neither_root_nor_terminal), locs.end());
+    util::sort(locs);
+    util::unique_in_place(locs);
 
-    mlocation_map head_count;
-    unsigned extra_cv_count = 0;
+    // Collect cables constituting each CV, maintaining a stack of CV
+    // proximal 'head' points, and recursing down branches in the morphology
+    // within each CV.
+
+    constexpr fvm_index_type no_parent = -1;
+    std::vector<std::pair<mlocation, fvm_index_type>> next_cv_head; // head loc, parent cv index
+    next_cv_head.emplace_back(mlocation{mnpos, 0}, no_parent);
```

Comment:
`{mnpos, 0}` instead of `{mnpos, 1}` ?

Reply:
Here, it doesn't matter. But if we move to allowing `{mnpos, 1}` to represent the more-proximal-than-root point, we should use that.

Reply:
Agreed, it doesn't matter here, but "parent is on the distal end of parent branch" feels like a good pattern to follow.